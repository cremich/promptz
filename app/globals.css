pre,
code {
  word-wrap: break-word !important;
  white-space: pre-wrap !important;
  white-space: -moz-pre-wrap !important;
  white-space: -pre-wrap !important;
  white-space: -o-pre-wrap !important;
  margin: 0;
  padding: 0;
}

code {
  background-color: #f3f3f7;
}

pre > code {
  display: inline-block;
  padding: 1em;
}
