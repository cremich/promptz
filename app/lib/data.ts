import {
  PromptCategory,
  QInterface,
  SdlcActivity,
} from "@/app/lib/definitions";

function convertEnumToTags(enumObj: Record<string, string>): string[] {
  return Object.entries(enumObj)
    .filter(([key]) => isNaN(Number(key)))
    .map(([_, value]) => value);
}

export function getAllTags(): string[] {
  const sdlcTags = convertEnumToTags(SdlcActivity);
  const categoryTags = convertEnumToTags(PromptCategory);
  const interfaceTags = convertEnumToTags(QInterface);
  return [...sdlcTags, ...categoryTags, ...interfaceTags];
}

export function getQInterfaceTags(): string[] {
  return convertEnumToTags(QInterface);
}

export function getCategoryTags(): string[] {
  return convertEnumToTags(PromptCategory);
}

export function getSdlcTags(): string[] {
  return convertEnumToTags(SdlcActivity);
}
