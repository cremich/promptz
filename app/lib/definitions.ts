import { z } from "zod";

export enum SdlcActivity {
  DEBUG = "Debugging",
  DEPLOY = "Deploy",
  DESIGN = "Design",
  DOCUMENT = "Documentation",
  ENHANCE = "Enhance",
  IMPLEMENT = "Implement",
  OPERATE = "Operate",
  OPTIMIZE = "Optimize",
  PATCH = "Patch Management",
  PLAN = "Plan",
  REFACTOR = "Refactoring",
  REQ = "Requirements",
  SECURITY = "Security",
  SUPPORT = "Support",
  TEST = "Test",
}

export enum PromptCategory {
  CHAT = "Chat",
  DEV_AGENT = "Dev Agent",
  DOC_AGENT = "Doc Agent",
  INLINE = "Inline",
  REVIEW_AGENT = "Review Agent",
  TEST_AGENT = "Test Agent",
  TRANSFORM = "Transform Agent",
  TRANSLATE = "Translate",
}

export enum QInterface {
  IDE = "IDE",
  CLI = "CLI",
  CONSOLE = "Management Console",
}

export type User = {
  id: string;
  username: string;
  displayName: string;
  guest: boolean;
};

export type Prompt = {
  id?: string;
  title?: string;
  description?: string;
  tags?: string[];
  instruction?: string;
  howto?: string;
  author?: string;
  authorId?: string;
  public?: boolean;
  slug?: string;
  sourceURL?: string;
  createdAt?: string;
  updatedAt?: string;
};

const ALLOWED_DOMAINS = [
  "github.com",
  "dev.to",
  "hashnode.com",
  "medium.com",
  "stackoverflow.com",
  "community.aws",
  "linkedin.com",
  "aws.amazon.com",
  "docs.aws.amazon.com",
  "amazon.com",
  "amazon.science",
  "huggingface.co",
  "kaggle.com",
  "paperswithcode.com",
  "readthedocs.io",
  "gitbook.io",
];

export const promptFormSchema = z.object({
  id: z.string().uuid().optional(),
  sourceURL: z
    .string()
    .max(2048, "URL must not exceed 2048 characters")
    .url()
    .regex(/^https:\/\/.+/, "Only HTTPS URLs are allowed")
    .refine((url) => {
      try {
        const domain = new URL(url).hostname;
        return ALLOWED_DOMAINS.some(
          (allowed) => domain === allowed || domain.endsWith(`.${allowed}`),
        );
      } catch {
        return false;
      }
    }, "Domain not allowed. Please use a supported content platform.")
    .optional()
    .or(z.literal("")),
  title: z
    .string()
    .trim()
    .max(100, "Title must be less than 100 characters")
    .min(3, "Title must be more than 3 characters"),
  description: z
    .string()
    .trim()
    .min(10, "Description must be more than 10 characters")
    .max(500, "Description must be less than 500 characters"),
  howto: z
    .string()
    .trim()
    .max(4000, "How to must be less than 4000 characters")
    .optional(),
  instruction: z
    .string()
    .trim()
    .min(10, "Prompt must be more than 10 characters")
    .max(4000, "Prompt must be less than 4000 characters"),
  tags: z.array(z.string()).optional(),
  public: z.boolean(),
});

// Validation schema for search and filter params
export const searchParamsSchema = z.object({
  query: z.string().optional(),
  sort: z.string().optional(),
  my: z.string().optional(),
  interface: z.union([z.string(), z.array(z.string())]).optional(),
  category: z.union([z.string(), z.array(z.string())]).optional(),
  sdlc: z.union([z.string(), z.array(z.string())]).optional(),
});
