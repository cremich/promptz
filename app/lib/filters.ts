export type FilterCondition = Record<string, unknown>;

// Utility function for case variations
export function getCaseVariations(text: string): string[] {
  return [
    text.toLowerCase(),
    text.toUpperCase(),
    text.charAt(0).toUpperCase() + text.slice(1).toLowerCase(),
  ];
}

// Handle text search conditions
export function buildTextSearchFilter(query: string): FilterCondition {
  const variations = getCaseVariations(query);
  return {
    or: variations.flatMap((variant) => [
      { name: { contains: variant } },
      { description: { contains: variant } },
    ]),
  };
}

// Handle tag-based filtering
export function buildTagFilter(params: string | string[]): FilterCondition[] {
  const tags = Array.isArray(params) ? params : [params];
  return tags.map((tag) => ({
    tags: { contains: tag },
  }));
}
