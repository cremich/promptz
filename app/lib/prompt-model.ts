import {
  idSchema,
  sourceURLSchema,
  titleSchema,
  descriptionSchema,
  tagSchema,
  publicSchema,
} from "@/lib/forms/schema-definitions";
import { z } from "zod";

export type Prompt = {
  id?: string;
  title?: string;
  description?: string;
  tags?: string[];
  instruction?: string;
  howto?: string;
  author?: string;
  authorId?: string;
  public?: boolean;
  slug?: string;
  sourceURL?: string;
  createdAt?: string;
  updatedAt?: string;
  copyCount?: number;
  starCount?: number;
  popularityScore?: number;
};
export const promptFormSchema = z.object({
  id: idSchema,
  sourceURL: sourceURLSchema,
  title: titleSchema,
  description: descriptionSchema,
  howto: z
    .string()
    .trim()
    .max(4000, "How to must be less than 4000 characters")
    .optional(),
  instruction: z
    .string()
    .trim()
    .min(10, "Prompt must be more than 10 characters")
    .max(4000, "Prompt must be less than 4000 characters"),
  tags: tagSchema,
  public: publicSchema,
}); // Validation schema for search and filter params

export const promptSearchParamsSchema = z.object({
  query: z.string().optional(),
  sort: z.string().optional(),
  tags: z.union([z.string(), z.array(z.string())]).optional(),
});
