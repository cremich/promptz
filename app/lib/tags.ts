import {
  PromptCategory,
  QInterface,
  SdlcActivity,
} from "@/app/lib/definitions";

function convertEnumToTags(enumObj: Record<string, string>): string[] {
  return Object.values(enumObj).filter((value) => typeof value === "string");
}

export function getQInterfaceTags(): string[] {
  return convertEnumToTags(QInterface);
}

export function getCategoryTags(): string[] {
  return convertEnumToTags(PromptCategory);
}

export function getSdlcTags(): string[] {
  return convertEnumToTags(SdlcActivity);
}
