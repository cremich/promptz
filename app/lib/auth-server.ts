import { cookies } from "next/headers";
import { NextRequest, NextResponse } from "next/server";
import { createServerRunner } from "@aws-amplify/adapter-nextjs";
import { fetchAuthSession, fetchUserAttributes } from "aws-amplify/auth/server";

import outputs from "@/amplify_outputs.json";
import { User } from "@/app/lib/definitions";

const { runWithAmplifyServerContext } = createServerRunner({
  config: outputs,
});

export async function fetchCurrentAuthUser(): Promise<User> {
  try {
    const currentUser = await runWithAmplifyServerContext({
      nextServerContext: { cookies },
      operation: (contextSpec) => fetchUserAttributes(contextSpec),
    });
    return { displayName: currentUser.preferred_username!, guest: false };
  } catch (error) {
    return { displayName: "", guest: true };
  }
}

export async function checkUserIsAuthenticated(
  request: NextRequest,
  response: NextResponse,
) {
  const authenticated = await runWithAmplifyServerContext({
    nextServerContext: { request, response },
    operation: async (contextSpec) => {
      try {
        const session = await fetchAuthSession(contextSpec, {});
        return session.tokens !== undefined;
      } catch {
        return false;
      }
    },
  });
  return authenticated;
}
