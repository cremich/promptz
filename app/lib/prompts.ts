// utils/amplify-utils.ts
import { cookies } from "next/headers";

import { createServerRunner } from "@aws-amplify/adapter-nextjs";
import { generateServerClientUsingCookies } from "@aws-amplify/adapter-nextjs/api";

import { type Schema } from "@/amplify/data/resource";
import outputs from "@/amplify_outputs.json";
import { Prompt } from "@/app/lib/definitions";

export const { runWithAmplifyServerContext } = createServerRunner({
  config: outputs,
});

export const appsync = generateServerClientUsingCookies<Schema>({
  config: outputs,
  cookies,
});

export async function fetchFeaturedPrompts(): Promise<Prompt[]> {
  console.log("fetched from server");
  const { data: prompts, errors } = await appsync.models.prompt.list({
    limit: 3,
  });

  if (errors && errors.length > 0) {
    throw new Error(errors[0].message);
  }

  const promptList = prompts.map((p) => {
    const tags = Object.entries({
      category: p.category,
      sdlc_phase: p.sdlc_phase,
      interface: p.interface,
    })
      .map(([_, value]) => value)
      .filter(Boolean);

    return {
      id: p.id,
      title: p.name,
      description: p.description,
      tags: tags,
      author: p.owner_username,
    } as Prompt;
  });
  return promptList;
}
