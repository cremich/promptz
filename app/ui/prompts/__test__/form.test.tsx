import { Prompt } from "@/app/lib/definitions";
import PromptForm from "@/app/ui/prompts/form";
import { describe, expect, test, jest } from "@jest/globals";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";

describe("Prompt Form", () => {
  const prompt: Prompt = {
    id: "12345",
    title: "Test Prompt",
  };

  test("renders the prompt form component unchanged", () => {
    const { container } = render(<PromptForm prompt={prompt} />);
    expect(container).toMatchSnapshot();
  });
});
