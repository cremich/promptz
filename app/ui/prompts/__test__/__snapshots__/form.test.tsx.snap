// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prompt Form renders the prompt form component unchanged 1`] = `
<div>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    class="space-y-8"
  >
    <input
      name="id"
      type="hidden"
      value="12345"
    />
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 flex items-center gap-2"
        for=":r0:-form-item"
      >
        <svg
          class="lucide lucide-tag w-4 h-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.586 2.586A2 2 0 0 0 11.172 2H4a2 2 0 0 0-2 2v7.172a2 2 0 0 0 .586 1.414l8.704 8.704a2.426 2.426 0 0 0 3.42 0l6.58-6.58a2.426 2.426 0 0 0 0-3.42z"
          />
          <circle
            cx="7.5"
            cy="7.5"
            fill="currentColor"
            r=".5"
          />
        </svg>
        Title
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-base ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 md:text-sm text-white placeholder-opacity-50"
        id=":r0:-form-item"
        name="title"
        placeholder="Enter a catchy title"
        value="Test Prompt"
      />
      <p
        class="text-sm text-muted-foreground"
        id=":r0:-form-item-description"
      >
        A clear and descriptive title for your prompt
      </p>
    </div>
    <div
      class="flex items-center gap-4"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
        type="submit"
      >
        <svg
          class="lucide lucide-send w-4 h-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14.536 21.686a.5.5 0 0 0 .937-.024l6.5-19a.496.496 0 0 0-.635-.635l-19 6.5a.5.5 0 0 0-.024.937l7.93 3.18a2 2 0 0 1 1.112 1.11z"
          />
          <path
            d="m21.854 2.147-10.94 10.939"
          />
        </svg>
        Update Prompt
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 px-4 py-2"
        type="button"
      >
        <svg
          class="lucide lucide-save w-4 h-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.2 3a2 2 0 0 1 1.4.6l3.8 3.8a2 2 0 0 1 .6 1.4V19a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2z"
          />
          <path
            d="M17 21v-7a1 1 0 0 0-1-1H8a1 1 0 0 0-1 1v7"
          />
          <path
            d="M7 3v4a1 1 0 0 0 1 1h7"
          />
        </svg>
        Save as Draft
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
        type="button"
      >
        Cancel
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-destructive text-destructive-foreground hover:bg-destructive/90 h-10 px-4 py-2 ml-auto"
        type="button"
      >
        <svg
          class="lucide lucide-trash2 w-4 h-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 6h18"
          />
          <path
            d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
          />
          <path
            d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
          />
          <line
            x1="10"
            x2="10"
            y1="11"
            y2="17"
          />
          <line
            x1="14"
            x2="14"
            y1="11"
            y2="17"
          />
        </svg>
        Delete Prompt
      </button>
    </div>
    <div />
  </form>
</div>
`;
